src/charm.py

```diff
@@ -86,6 +87,32 @@ class COSProxyCharm(CharmBase):
 
     _stored = StoredState()
 
+    mandatory_relation_pairs = {
+        "dashboards": [  # must be paired with:
+            {"cos-agent"},  # or
+            {"downstream-grafana-dashboard"},
+        ],
+        "filebeat": [  # must be paired with:
+            {"downstream-logging"},
+        ],
+        "monitors": [  # (nrpe) must be paired with:
+            {"cos-agent"},  # or
+            {"downstream-prometheus-scrape"},
+        ],
+        "prometheus-target": [  # must be paired with:
+            {"cos-agent"},  # or
+            {"downstream-prometheus-scrape"},
+        ],
+        # "prometheus-rules": [  # must be paired with:
+        #     {"cos-agent"},  # or
+        #     {"downstream-prometheus-scrape"},
+        # ],
+        # "prometheus": [  # must be paired with:
+        #     {"cos-agent"},  # or
+        #     {"downstream-prometheus-scrape"},
+        # ],
```

Comment:
Seems like these also have mandatory pairs, but we don't track them with a stored state yet.
Could add this in as a comment or remove entirely. Wdyt?

Reply:
@simskij could you please review these combos?

Reply:
LGTM, however I have some doubts which I'll write in a comment below